src/sceneGraph/index.js

```diff
@@ -0,0 +1,596 @@
+     * @param {boolean} preserveWorldPosition - if true, will recompute a new localMatrix so that its worldMatrix will
+     * stay the same. if false, will keep the same localMatrix and compute a new worldMatrix
+     */
+    function changeParent(sceneNode, newParentId, preserveWorldPosition) {
+        let parentNode = getSceneNodeById(newParentId);
+        if (!parentNode) {
+            console.warn('can\'t find the parent SceneNode to add this to');
+        }
+        // check if the parent needs all of its children to be added to a rotateX node inside of it instead
+        parentNode = getNodeOrRotateXChild(parentNode);
+        sceneNode.setParent(parentNode);
+
+        if (preserveWorldPosition) {
+            let desiredWorldMatrix = sceneNode.worldMatrix;
+            let newParentWorldMatrix = sceneNode.parent.worldMatrix;
+            let requiredLocalMatrix = [];
+            utils.multiplyMatrix(desiredWorldMatrix, utils.invertMatrix(newParentWorldMatrix), requiredLocalMatrix);
+            let transform = sceneNode.getTransformMatrix();
+            let inverseTransform = utils.invertMatrix(transform);
+            let untransformed = [];
+            utils.multiplyMatrix(inverseTransform, requiredLocalMatrix, untransformed);
+            sceneNode.setLocalMatrix(untransformed);
+        }
+
+        sceneNode.flagForRecompute();
+    }
+
+    function changeId(sceneNode, newId) {
+        let oldId = sceneNode.id;
+        sceneNode.id = newId;
+
+        sceneGraph[sceneNode.id] = sceneNode;
+        delete sceneGraph[oldId];
+
+        delete relativeToCamera[oldId];
+        delete finalCSSMatrices[oldId];
+        delete finalCSSMatricesWithoutTransform[oldId];
+
+        // this.dirty = true;
+        // this.flagAsDirty(); // this will make sure computations get computed/cached with new ID
+        sceneNode.flagForRecompute(); // TODO: might not be necessary? just say e.g. data[new] = data[old]; delete data[old];
+    }
+
+    /************ Private Functions ************/
+    function addRotateX(sceneNodeObject, objectId, groundPlaneVariation) {
+        let sceneNodeRotateX;
+        let thisNodeId = objectId + 'rotateX';
+        if (typeof sceneGraph[thisNodeId] !== 'undefined') {
+            console.warn('trying to add duplicate rotateX to scene graph');
+            sceneNodeRotateX = sceneGraph[thisNodeId];
+        } else {
+            sceneNodeRotateX = new SceneNode(thisNodeId);
+            // let _object = realityEditor.getObject(objectId);
+            sceneGraph[thisNodeId] = sceneNodeRotateX;
+        }
+
+        sceneNodeRotateX.setParent(sceneNodeObject);
+        console.log('SceneGraph: added rotateX to object ' + objectId);
+
+        // image target objects require one coordinate system rotation. ground plane requires another.
+        if (groundPlaneVariation) {
+            sceneNodeRotateX.setLocalMatrix( makeGroundPlaneRotationX(-(Math.PI/2)) );
```

Comment:
nit: Parentheses are a little weird here

Reply:
Thank you for such a thorough review!